plugins/inputs/intel_powerstat/services/rapl.go

```diff
@@ -0,0 +1,274 @@
+// +build linux
+
+package services
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs/intel_powerstat/data"
+)
+
+// Paths for Intel RAPL.
+const (
+	IntelRaplPath                = "/sys/devices/virtual/powercap/intel-rapl"
+	IntelRaplSocketPartialPath   = "%s/intel-rapl:%s"
+	EnergyUjPartialPath          = "%s/energy_uj"
+	MaxRangeUjPartialPath        = "%s/max_energy_range_uj"
+	MaxPowerUwPartialPath        = "%s/constraint_0_max_power_uw"
+	IntelRaplDramPartialPath     = "%s/intel-rapl:%s/%s"
+	IntelRaplDramNamePartialPath = "%s/name"
+)
+
+// RaplService is responsible for interactions with RAPL.
+type RaplService interface {
+	InitializeRaplData()
+	GetConstraintMaxPower(socketID string) (float64, error)
+	RetrieveAndCalculateData(socketID string) error
+	GetSocketIDs() []string
+	GetCurrentPackagePowerConsumption(socketID string) float64
+	GetCurrentDramPowerConsumption(socketID string) float64
+	GetDramMaxEnergyRangeUj(socketID string) (float64, error)
+	GetMaxEnergyRangeUj(socketID string) (float64, error)
+}
+
+// RaplServiceImpl is implementation of RaplService.
+type RaplServiceImpl struct {
+	log          telegraf.Logger
```

Comment:
Again I think you should not rely on telegraf here. Remove the logger and return errors where needed.

Reply:
This is integral part of Telegraf plugin. Abstraction layer simplified.